cmd/export/cluster.go

```diff
@@ -0,0 +1,243 @@
+
+func NewClusterScopeHandler() *ClusterScopeHandler {
+	clusterScopeHandler := &ClusterScopeHandler{}
+
+	return clusterScopeHandler
+}
+
+func isClusterScopedResource(apiGroup string, kind string) bool {
+	for _, admitted := range admittedClusterScopeResources {
+		if admitted.Kind == kind && admitted.APIgroup == apiGroup {
+			return true
+		}
+	}
+	return false
+}
+
+func (c *ClusterScopeHandler) filterRbacResources(resources []*groupResource, log logrus.FieldLogger) []*groupResource {
+	log.Debug("Looking for ServiceAccount resources")
+
+	handler := NewClusterScopedRbacHandler(log)
+	var filteredResources []*groupResource
+	for _, r := range resources {
+		kind := r.APIResource.Kind
+		if kind == "ServiceAccount" {
+			for _, obj := range r.objects.Items {
+				log.Debugf("Adding ServiceAccount %s in namespace %s", obj.GetName(), obj.GetNamespace())
+				handler.serviceAccounts = append(handler.serviceAccounts, obj)
+			}
+		}
+		if isClusterScopedResource(r.APIGroup, kind) {
+			log.Debugf("Adding %d Cluster resource of type %s", len(r.objects.Items), kind)
+			handler.clusterResources[kind] = r
+		} else {
+			filteredResources = append(filteredResources, r)
+		}
+	}
+
+	for _, k := range admittedClusterScopeResources {
+		filtered, ok := handler.filteredResourcesOfKind(k)
+		if ok && len(filtered.objects.Items) > 0 {
+			filteredResources = append(filteredResources, filtered)
+		}
+	}
+
+	return filteredResources
+}
+
+type ClusterScopedRbacHandler struct {
+	log              logrus.FieldLogger
+	readyToFilter    bool
+	serviceAccounts  []unstructured.Unstructured
+	clusterResources map[string]*groupResource
+
+	filteredClusterRoleBindings *groupResource
+}
+
+func NewClusterScopedRbacHandler(log logrus.FieldLogger) *ClusterScopedRbacHandler {
+	handler := &ClusterScopedRbacHandler{log: log, readyToFilter: false}
+	handler.clusterResources = make(map[string]*groupResource)
+	return handler
+}
+
+func (c *ClusterScopedRbacHandler) prepareForFiltering() {
+	c.readyToFilter = true
+	c.log.Infof("Preparing matching ClusterRoleBindings")
+
+	clusterRoleBindings, ok := c.clusterResources["ClusterRoleBinding"]
+	if ok {
+		c.filteredClusterRoleBindings = &groupResource{
+			APIGroup:        clusterRoleBindings.APIGroup,
+			APIVersion:      clusterRoleBindings.APIVersion,
+			APIGroupVersion: clusterRoleBindings.APIGroupVersion,
+			APIResource:     clusterRoleBindings.APIResource,
+		}
+		filteredClusterRoleBindings := unstructured.UnstructuredList{Items: []unstructured.Unstructured{}}
+		for _, crb := range clusterRoleBindings.objects.Items {
+			if c.acceptClusterRoleBinding(crb) {
+				filteredClusterRoleBindings.Items = append(filteredClusterRoleBindings.Items, crb)
+				c.log.Infof("Found matching %s %s", crb.GetKind(), crb.GetName())
+			}
+		}
+		c.filteredClusterRoleBindings.objects = &filteredClusterRoleBindings
+	}
+}
+
+func (c *ClusterScopedRbacHandler) filteredResourcesOfKind(resource admittedResource) (*groupResource, bool) {
```

Comment:
This filter logic, to me, seems more appropriate for the `transform` step.

Reply:
Hey @djzager, I perfectly agree with you that this seems to be a `transform` step, and you almost anticipated some of the reasons why I haven't followed this path (e.g., complexity and more repos impacted), but at the moment this can't be done without adding any context related to the other exported resources, because the PluginRun interface works only on individual resources, but the decision whether to accept or whiteout an RBAC resource depends also on other resources (e.g., `is there any ServiceAccount matching this ClusterRoleBinding`?). In brief, we should pass some kind of "context" to the plugin execution which, in turn, for each and every RBAC resource, should unmarshal all the files in all the exported namespaces to look for these matches.
I'm totally worried by performance, as you can imagine, this is why the initial implementation, even if it broke the usual 3-steps approach, looked more promising.
WDYT?

Reply:
Fair point. I believe the right answer for us in this scenario would be to articulate why we can't put this in transform plugin(s) via GitHub issue and proceed with this PR as it is currently implemented.

Reply:
I am plus one, a follow up issue would be great